pkg/client/cache/fifo.go

```diff
@@ -241,6 +241,17 @@ func (f *FIFO) Replace(list []interface{}, resourceVersion string) error {
 	return nil
 }
 
+// Resync will touch all objects to put them into the processing queue
+func (f *FIFO) Resync() error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	for id := range f.items {
+		f.queue = append(f.queue, id)
+	}
+	f.cond.Broadcast()
```

Comment:
So the map of `items` contains a key for `id`. The queue should already
contain `id` as well since it's an ordered representation
(super-collection? not quite a set of unique things) of the keys in the
`items` map. How does re-adding the ids to the end of the queue (where they
already exist, albeit earlier in the ordering) serve to provide resync
functionality? When an item is popped from the FIFO its entry is removed
from the map. Which means that its really no longer queued because unless
there's another (future) call to Add or Update, the id won't be present the
next time Pop is called for that id .. so how does Resync accomplish
anything here? Is this just to give priority processing to ids that FIFO
already knows about in the event of some "resync"-related event?

On Tue, May 10, 2016 at 12:16 PM, Timothy St. Clair <
notifications@github.com> wrote:

> In pkg/client/cache/fifo.go
> <https://github.com/kubernetes/kubernetes/pull/24142#discussion_r62702817>
> :
>
> > @@ -241,6 +241,17 @@ func (f *FIFO) Replace(list []interface{}, resourceVersion string) error {
> >  	return nil
> >  }
> >
> > +// Resync will touch all objects to put them into the processing queue
> > +func (f *FIFO) Resync() error {
> > +	f.lock.Lock()
> > +	defer f.lock.Unlock()
> > +	for id := range f.items {
> > +		f.queue = append(f.queue, id)
> > +	}
> > +	f.cond.Broadcast()
>
> This is how the resync is done in kubernetes.
>
> The resync will literally walk the items and force the higher order code
> walk through periodically to ensure the cluster state is up to date. In
> many cases it's redundant, but it's done on purpose.
>
> The purpose is so the controllers continually attempt to validate the
> state of the universe ~= declared state.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24142/files/5168833913b27499ec59eec2ad889a35f711d6d8#r62702817>
>